src/compiler/crystal/tools/typer.cr

```diff
@@ -0,0 +1,612 @@
+          type.def_instances.each do |_, def_instance|
+            next unless accepted_defs.keys.includes?(def_instance.location.to_s)
+
+            ret[def_instance.location.to_s] << def_instance
+          end
+        end
+
+        # Check for class 'self.def's
+        metaclass = type.metaclass
+        if metaclass.responds_to?(:def_instances)
+          metaclass.def_instances.each do |_, def_instance|
+            next unless accepted_defs.keys.includes?(def_instance.location.to_s)
+
+            ret[def_instance.location.to_s] << def_instance
+          end
+        end
+      end
+
+      # Now remove all overridden methods
+      overridden_method_locations.each do |child_loc, ancestor_loc|
+        if ret.delete(child_loc)
+          @warnings << "Not adding type restrictions to definition at #{child_loc} as it overrides definition #{ancestor_loc}"
+        end
+      end
+
+      ret
+    end
+
+    private def def_overrides_parent_def(type) : Hash(String, String)
+      overriden_locations = {} of String => String
+      type.defs.try &.each_value do |defs_with_metadata|
+        defs_with_metadata.each do |def_with_metadata|
+          next if def_with_metadata.def.location.to_s.starts_with?("expanded macro:") || def_with_metadata.def.name == "initialize"
+          type.ancestors.each do |ancestor|
+            ancestor_defs_with_metadata = ancestor.defs.try &.[def_with_metadata.def.name]?
+            ancestor_defs_with_metadata.try &.each do |ancestor_def_with_metadata|
+              next if ancestor_def_with_metadata.def.location.to_s.starts_with?("expanded macro:")
+              found_def_with_same_name = true
+
+              if def_with_metadata.compare_strictness(ancestor_def_with_metadata, self_owner: type, other_owner: ancestor) == 0
+                overriden_locations[def_with_metadata.def.location.to_s] = ancestor_def_with_metadata.def.location.to_s
+                overriden_locations[ancestor_def_with_metadata.def.location.to_s] = def_with_metadata.def.location.to_s
+              end
+            end
+          end
+        end
+      end
+      overriden_locations
+    end
+
+    # Given an 'arg', return its type that's good for printing (VirtualTypes suffix themselves with a '+')
+    private def resolve_type(arg)
+      t = arg.type
+      t.is_a?(Crystal::VirtualType) ? t.base_type : t
+    end
+
+    # Strip out any NoReturns, or Procs that point to them (maybe all generics? Start with procs)
+    private def filter_no_return(types)
+      compacted_types = types.to_a.reject! do |type|
+        type.no_return? || (type.is_a?(Crystal::ProcInstanceType) && type.as(Crystal::ProcInstanceType).return_type.no_return?)
+      end
+
+      compacted_types << program.nil if compacted_types.empty?
+      compacted_types
+    end
+
+    # Generates a map of (parsed) Def#location => Signature for that Def
+    private def init_signatures(accepted_defs : Hash(String, Crystal::Def)) : Hash(String, Signature)
```

Comment:
Workhorse method that generates the hash of location strings => signature for the def at that location